src/Orleans.Core/Runtime/CallbackData.cs

```diff
@@ -1,108 +1,64 @@
-            Action<Message> unregisterDelegate,
-            MessagingOptions messagingOptions,
-            ILogger logger,
-            ILogger timerLogger)
+            Message msg)
         {
-            // We are never called without a callback func, but best to double check.
-            if (callback == null) throw new ArgumentNullException(nameof(callback));
-            // We are never called without a resend func, but best to double check.
-            if (resendFunc == null) throw new ArgumentNullException(nameof(resendFunc));
-            this.logger = logger;
-            this.callback = callback;
-            this.resendFunc = resendFunc;
-            context = ctx;
-            Message = msg;
-            unregister = unregisterDelegate;
-            alreadyFired = false;
-            this.messagingOptions = messagingOptions;
+            this.shared = shared;
+            this.context = ctx;
+            this.Message = msg;
             this.TransactionInfo = TransactionContext.GetTransactionInfo();
-            this.timerLogger = timerLogger;
+            this.durationTimestamp = Stopwatch.GetTimestamp();
         }
 
-        /// <summary>
-        /// Start this callback timer
-        /// </summary>
-        /// <param name="time">Timeout time</param>
-        public void StartTimer(TimeSpan time)
-        {
-            if (time < TimeSpan.Zero)
-                throw new ArgumentOutOfRangeException(nameof(time), "The timeout parameter is negative.");
-            timeout = time;
-            if (StatisticsCollector.CollectApplicationRequestsStats)
-            {
-                timeSinceIssued = TimeIntervalFactory.CreateTimeInterval(true);
-                timeSinceIssued.Start();
-            }
+        public ITransactionInfo TransactionInfo { get; set; }
 
-            TimeSpan firstPeriod = timeout;
-            TimeSpan repeatPeriod = Constants.INFINITE_TIMESPAN; // Single timeout period --> No repeat
-            if (messagingOptions.ResendOnTimeout && messagingOptions.MaxResendCount > 0)
-            {
-                firstPeriod = repeatPeriod = timeout.Divide(messagingOptions.MaxResendCount + 1);
-            }
-            // Start time running
-            DisposeTimer();
-            timer = new SafeTimer(this.timerLogger, TimeoutCallback, null, firstPeriod, repeatPeriod);
+        public Message Message { get; set; } // might hold metadata used by response pipeline
+
+        public bool IsCompleted => this.alreadyFired;
 
+        public bool IsExpired(long currentTimestamp)
+        {
+            return currentTimestamp - this.durationTimestamp > this.shared.ResponseTimeoutStopwatchTicks;
         }
 
-        private void TimeoutCallback(object obj)
+        public TimeSpan GetCallDuration()
         {
-            OnTimeout();
+            // A positive timestamp value indicates the start time of an ongoing call,
+            // a negative value indicates the negative total duration of a completed call.
```

Comment:
Would it be more clear to introduce nullable `totalDuration` field and `requestStartTimestamp`, so that implicit meaning of negative values could be avoided?

Reply:
Maybe we should abstract it into a struct instead, eg `ValueStopwatch`, what do you think?

Reply:
I don't see another places where such struct could be used, and `ValueStopwatch` name does not fully reflect its purpose so it might be better not to add additional entity, when it can be avoided.

Reply:
And public `startTimestamp` property. Yes, that makes sense.